src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -0,0 +1,345 @@
+
+        /// <summary>
+        /// Parses an HTTP form body.
+        /// </summary>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/>.</param>
+        /// <returns>The collection containing the parsed HTTP form body.</returns>
+        public async Task<Dictionary<string, StringValues>> ReadFormAsync(CancellationToken cancellationToken = default)
+        {
+            KeyValueAccumulator accumulator = default;
+            while (true)
+            {
+                ReadResult readResult;
+                if (!_pipeReader.TryRead(out readResult))
+                {
+                    readResult = await _pipeReader.ReadAsync(cancellationToken);
+                }
+
+                var buffer = readResult.Buffer;
+
+                if (!buffer.IsEmpty)
+                {
+                    TryParseFormValues(ref buffer, ref accumulator, readResult.IsCompleted, _encoding, KeyLengthLimit, ValueLengthLimit, ValueCountLimit);
+                }
+
+                if (readResult.IsCompleted)
+                {
+                    break;
+                }
+
+                _pipeReader.AdvanceTo(buffer.Start, buffer.End);
+            }
+
+            return accumulator.GetResults();
+        }
+
+        // Used for testing/perf.
+        internal static void TryParseFormValues(ref ReadOnlySequence<byte> buffer, ref KeyValueAccumulator accumulator, bool isFinalBlock)
+        {
+            TryParseFormValues(ref buffer, ref accumulator, isFinalBlock, Encoding.UTF8);
+        }
+
+        internal static void TryParseFormValues(ref ReadOnlySequence<byte> buffer, ref KeyValueAccumulator accumulator, bool isFinalBlock, Encoding encoding)
+        {
+            TryParseFormValues(ref buffer, ref accumulator, isFinalBlock, encoding, DefaultKeyLengthLimit, DefaultValueLengthLimit, DefaultValueCountLimit);
+        }
+
+        internal static void TryParseFormValues(
+            ref ReadOnlySequence<byte> buffer,
+            ref KeyValueAccumulator accumulator,
+            bool isFinalBlock,
+            Encoding encoding,
+            int keyLengthLimit,
+            int valueLengthLimit,
+            int valueCountLimit)
+        {
+            // We will eventually be comparing against equals and ampersand, so get the encoded
+            // versions of them. For Utf8/ASCII, these are static, so getting the encoding is zero allocation.
+            var equalsEncoded = GetEqualsForEncoding(encoding);
+            var andEncoded = GetAndForEncoding(encoding);
+
+            if (buffer.IsSingleSegment)
+            {
+                TryParseFormValuesFast(buffer.First.Span,
+                    ref accumulator,
+                    isFinalBlock,
+                    encoding,
+                    keyLengthLimit,
+                    valueLengthLimit,
+                    valueCountLimit,
+                    equalsEncoded,
+                    andEncoded,
+                    out var consumed);
```

Comment:
Why out consumed only to slice? Can't you pass the buffer in as ref and slice? Or is that slower?

Reply:
You need to update the sequence so there’s no point slicing the span.